### PR TITLE
increase block-relay-only connections 

### PR DESCRIPTION
**Pull request by amitiuttarwar**

### CODE REVIEW

**Comment by amitiuttarwar** [comment]
- addressed martin's two review comments: updated the sed command in the scripted diff commit message & added the `std::max(0,...)` guard for the inbound max

**Comment by amitiuttarwar** [comment]
two updates to the scripted diff 
* added the `m_max_outbound` -> `m_max_automatic_outbound` rename to scripted diff using `/1` trick 
* changed the script slightly to be less confusing around `nMaxConnections` -> `m_max_automatic_connections`

File: src/net.h [hunk 1]
```diff
@@ -75,7 +75,9 @@ static const int MAX_FEELER_CONNECTIONS = 1;
 /** -listen default */
 static const bool DEFAULT_LISTEN = true;
 /** The maximum number of peer connections to maintain. */
-static const unsigned int DEFAULT_MAX_PEER_CONNECTIONS = 125;
+static const unsigned int DEFAULT_MAX_PEER_CONNECTIONS{200};
+/** Percentage of inbound connection slots that tx-relaying peers can use */
+static const float FULL_RELAY_INBOUND_PCT{0.6};
```
Comment (mzumsande) [inline src/net.h]: No idea why windows CI runs on your repo, but it's nice. It complains about the float - will change it to double later.

File: src/net.h [hunk 2]
```diff
@@ -824,7 +825,7 @@ class CConnman
         m_max_outbound_block_relay = std::min(MAX_BLOCK_RELAY_ONLY_CONNECTIONS, m_max_automatic_connections - m_max_outbound_full_relay);
         m_max_automatic_outbound = m_max_outbound_full_relay + m_max_outbound_block_relay + m_max_feeler;
         m_max_inbound = std::max(0, m_max_automatic_connections - m_max_automatic_outbound);
-        m_max_inbound_full_relay = std::max(0, static_cast<int>(FULL_RELAY_INBOUND_PCT * m_max_inbound));
+        m_max_inbound_full_relay = std::min(1, std::max(0, static_cast<int>(connOptions.m_full_relay_inbound_ratio * m_max_inbound)));
```
Comment (mzumsande) [inline src/net.h:828]: the added `min(1, ...)` seems incorrect (and breaks the test). It means we can't ever have more than one FR inbound?

File: src/net.h (hunk 1 above)
Comment (mzumsande) [reply src/net.h]: actually changed it to int because it's a percentage.

**Comment by mzumsande** [comment]
So I made a few changes:
- fixed `p2p_eviction.py` test (in a separate commit for easier review, will squash later)
- changed `FULL_RELAY_INBOUND_PCT` to an `int` (because it's a percentage)
- in commit WIP: update so ratio is init param: clamped `m_full_relay_inbound_ratio` it to a range between 0 and 100 in init (and removed the std::min in `CConnman::init` for that). That way it works and the tests succeed - feel free to improve this, e.g. by error messages if it's out of bounds, or other solutions you had in mind.
- added a commit to handle the bloom filter logic (+test): It wasn't possible to reuse the code from `CreateNodeFromAcceptedSocket` like I planned, because in that situation we haven't connected the new peer yet (and therefore need to check if we are at the limit) and also must check for total peer count, whereas in net_processing, the peer is already connected, and we just have to check after changing tx_relay if the tx-relay count is exceeded. The way I implemented is to first try the normal eviction logic, and if that fails (all peers are protected) we evict the peer that just sent us a filter.

File: src/net.cpp [hunk 3]
```diff
@@ -1698,6 +1698,23 @@ int CConnman::GetExtraBlockRelayCount() const
     return std::max(block_relay_peers - m_max_outbound_block_relay, 0);
 }
 
+bool CConnman::EvictTxPeerIfFull()
+{
+    int tx_inbound_peers{0};
```
Comment (amitiuttarwar) [inline src/net.cpp:1703]: can this function return early if `fListen` is false?

File: src/net.h [hunk 4]
```diff
@@ -927,6 +927,13 @@ class CConnman
     // Count the number of block-relay-only peers we have over our limit.
     int GetExtraBlockRelayCount() const;
 
+    /**
+     * If we are full, attempt to evict a tx-relaying peer
```
Comment (amitiuttarwar) [inline src/net.h]: maybe: If we are at capacity for inbound tx-relay peers, attempt to evict one.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -4770,6 +4770,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             }
             pfrom.m_bloom_filter_loaded = true;
             pfrom.m_relays_txs = true;
+            if (!m_connman.EvictTxPeerIfFull()) {
+                // We don't have room for another tx-relay peer, disconnect
+                LogPrint(BCLog::NET, "filterload received, but no capacity for tx-relay and no other peer to evict. disconnecing peer=%d\n", pfrom.GetId());
```
Comment (amitiuttarwar) [inline src/net_processing.cpp]: typo: disconnecting

**Comment by amitiuttarwar** [comment]
pushed changes to clean up the init param commit. small things like changed the variables to match the percent/int pattern, rewrote commit message, etc.

File: src/net.cpp (hunk 3 above)
Comment (mzumsande) [reply src/net.cpp:1703]: yes - I think we can just not call it in `net_processing` if the bloom peer isn't an inbound connection - will push an update for that.

File: src/init.cpp [hunk 6]
```diff
@@ -1727,6 +1728,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     CConnman::Options connOptions;
     connOptions.nLocalServices = nLocalServices;
     connOptions.m_max_automatic_connections = nMaxConnections;
+    connOptions.m_full_relay_inbound_percent = std::clamp<int>(args.GetIntArg("-inboundrelaypercent", DEFAULT_FULL_RELAY_INBOUND_PCT), 0, 100) / 100.0;
```
Comment (mzumsande) [inline src/init.cpp]: now that it's renamed from ratio to percent. we should keep it an int, and do the division with 100 just in connman init.

File: src/net.h (hunk 4 above)
Comment (mzumsande) [reply src/net.h]: yep, that's better!

File: src/init.cpp (hunk 6 above)
Comment (mzumsande) [reply src/init.cpp]: moved the division.

**Comment by mzumsande** [comment]
[56d9caf ](https://github.com/amitiuttarwar/bitcoin/commit/56d9cafe8b36125b8459e738f46d6d76fc87759e)to [460c974](https://github.com/amitiuttarwar/bitcoin/commit/460c974a803f01eedb753e10ad76bfc05ef7eb59) : addressed feedback, also pushed for CI. Will push once more at least.

File: src/net.h [hunk 7]
```diff
@@ -1197,19 +1197,25 @@ class CConnman
 
     std::unique_ptr<CSemaphore> semOutbound;
     std::unique_ptr<CSemaphore> semAddnode;
-    int m_max_automatic_connections;
-
-    // How many full-relay (tx, block, addr) outbound peers we want
```
Comment (mzumsande) [inline src/net.h:1207]: did you drop the doc for  `m_max_outbound_full_relay` and `m_max_outbound_block_relay` on purpose?

**Comment by mzumsande** [comment]
some more small changes:
commit "net: add m_max_inbound to connman"
expanded commit message, dropped "refactor" from it because of the added `max()`

commit "doc: Cleanup documentation around connection limit maximums"
added to m_max_automatic_connections the info that it includes inbounds

commit "net: increase inbound capacity for block-relay-only connections"
updated commit message

commit "test: add functional test for inbound maxconnection limits"
fixed possible timeout, changed doc a bit

**Pull request closed without merging**